crates/viewer/re_space_view_graph/src/layout/provider.rs

```diff
@@ -27,115 +27,88 @@ impl<'a> From<&'a NodeTemplate> for fj::Node {
             request,
         }
     }
 
-    pub fn init(&self) -> Layout {
-        let positions = self.simulation.positions().collect::<Vec<_>>();
-        let mut extents = ahash::HashMap::default();
-
-        for graph in self.request.graphs.values() {
-            for (id, node) in &graph.nodes {
-                let i = self.node_index[id];
-                let [x, y] = positions[i];
-                let pos = Pos2::new(x as f32, y as f32);
-                extents.insert(*id, Rect::from_center_size(pos, node.size));
+    pub fn new_with_previous(request: LayoutRequest, layout: &Layout) -> Self {
+        let nodes = request.all_nodes().map(|(id, v)| {
+            if let Some(rect) = layout.get_node(&id) {
+                let pos = rect.center();
+                fj::Node::from(v).position(pos.x as f64, pos.y as f64)
+            } else {
+                fj::Node::from(v)
             }
-        }
+        });
+        let edges = considered_edges(&request);
+
+        // TODO(grtlr): Currently we guesstimate good forces. Eventually these should be exposed as blueprints.
+        let simulation = fj::SimulationBuilder::default()
+            .with_alpha_decay(0.01) // TODO(grtlr): slows down the simulation for demo
+            .build(nodes)
+            .add_force("link", fj::Link::new(edges).distance(50.0).iterations(2))
+            .add_force("charge", fj::ManyBody::new())
+            // TODO(grtlr): This is a small stop-gap until we have blueprints to prevent nodes from flying away.
+            .add_force("x", fj::PositionX::new().strength(0.01))
+            .add_force("y", fj::PositionY::new().strength(0.01));
 
-        Layout {
-            nodes: extents,
-            // Without any real node positions, we probably don't want to draw edges either.
-            edges: ahash::HashMap::default(),
-            entities: Vec::new(),
+        Self {
+            simulation,
+            request,
         }
     }
 
-    /// Returns `true` if finished.
-    pub fn tick(&mut self, layout: &mut Layout) -> bool {
-        self.simulation.tick(1);
-
+    fn layout(&self) -> Layout {
+        // We make use of the fact here that the simulation is stable, i.e. the
+        // order of the nodes is the same as in the request.
+        let mut i = 0..;
         let positions = self.simulation.positions().collect::<Vec<_>>();
 
-        // We clear all unnecessary data from the previous layout, but keep its space allocated.
-        layout.entities.clear();
-        layout.edges.clear();
+        let mut layout = Layout::empty();
 
         for (entity, graph) in &self.request.graphs {
             let mut current_rect = Rect::NOTHING;
 
-            for node in graph.nodes.keys() {
-                let extent = layout.nodes.get_mut(node).expect("node has to be present");
-                let i = self.node_index[node];
+            for (node, template) in &graph.nodes {
+                let i = i.next().expect("can never happen");
```

Comment:
is that a fancy `i += 1`? :)

I wonder if there is a way to rewrite this logic a little bit. Can't you `.join()` `positions` with `graph.nodes` in the `for` statement directly?

Reply:
Oh, I forgot about slices—that's a good idea!